### PR TITLE
Make -lowmem optional and prefer building d_do_test with the host compiler

### PR DESCRIPTION
**Pull request by wilzbach**
CC @MoonlightSentinel

### CODE REVIEW

File: test/tools/sanitize_json.d [hunk 1]
```diff
@@ -13,6 +13,24 @@ import std.stdio;
...
+    alias JSON_TYPE_UINTEGER = JSON_TYPE.UINTEGER;
+} else {
+    alias JSON_TYPE_NULL = JSONType.null_;
+    alias JSON_TYPE_OBJECT = JSONType.object;
+    alias JSON_TYPE_STRING = JSONType.string;
+    alias JSON_TYPE_ARRAY = JSONType.array;
+    alias JSON_TYPE_INTEGER = JSONType.integer;
+    alias JSON_TYPE_UINTEGER = JSONType.uinteger;
+}
+
```
Comment (Geod24) [inline test/tools/sanitize_json.d:33]: Why not the other way around ?
```D
// JSONType has been introduced in 2.082
static if (__VERSION__ <= 2081) {
    enum JSONType : JSON_TYPE
    {
        null_ = JSON_TYPE.NULL;
        // ...
    }
}
```

This way you only have an import in the `else` and when we update the base version it's only removing code.

**Comment by MoonlightSentinel** [comment]
Interesting ... wasn't `-lowmem` introduced because of an OOM on the auto-tester or was it a different CI?

File: test/Makefile [hunk 2]
```diff
@@ -108,16 +108,17 @@ ifeq ($(HOST_DMD),)
 endif
 
 # Required version for -lowmem
-MIN_VERSION = v2.086.0
+LOW_MEM_MIN_VERSION = v2.086.0
 VERSION = $(filter v2.%, $(shell $(HOST_DMD) --version 2>/dev/null))
+RUN_FLAGS = -g -i -Itools -version=NoMain
 
 ifeq ($(VERSION),)
-    # dmd was not found in $PATH
```
Comment (MoonlightSentinel) [inline test/Makefile:115]: Why? Ditto for the other comments?

File: test/tools/d_do_test.d [hunk 3]
```diff
@@ -933,17 +933,17 @@ bool compareOutput(string output, string refoutput, const ref EnvData envData)
         foreach (const chunk; refparts[0].splitter('|'))
         {
             // ( <predicate> , "=", <content> )
-            const conditional = chunk.findSplit("=");
+            const searchResult = chunk.findSplit("=");
```
Comment (MoonlightSentinel) [inline test/tools/d_do_test.d:936]: This seems unrelated?

File: test/tools/sanitize_json.d (hunk 1 above)
Comment (wilzbach) [reply test/tools/sanitize_json.d:33]: Comparison of different enum types (std.json defines its `JsonType` enum).

**Comment by wilzbach** [comment]
> Interesting ... wasn't `-lowmem` introduced because of an OOM on the auto-tester or was it a different CI?

Yeah, very good point - https://github.com/dlang/dmd/commit/f9a4e897f80e35d20d49f9a72844cf4c0ecb29eb

File: test/Makefile (hunk 2 above)
Comment (wilzbach) [reply test/Makefile:115]: No reason. Sorry. Fixed.

File: test/tools/d_do_test.d (hunk 3 above)
Comment (wilzbach) [reply test/tools/d_do_test.d:936]: Yes, but it was a small change and to me a better name. Happy to change back if you're not convinced.

**Comment by wilzbach** [comment]
> > Interesting ... wasn't `-lowmem` introduced because of an OOM on the auto-tester or was it a different CI?
> 
> Yeah, very good point - [f9a4e89](https://github.com/dlang/dmd/commit/f9a4e897f80e35d20d49f9a72844cf4c0ecb29eb)

It's still the case:

image

File: test/tools/d_do_test.d (hunk 3 above)
Comment (MoonlightSentinel) [reply test/tools/d_do_test.d:936]: No strong feelings about either name but PRs / commits should not contain unrelated changes.

File: test/tools/sanitize_json.d (hunk 1 above)
Comment (Geod24) [reply test/tools/sanitize_json.d:33]: Erg that's really stupid. But again, `enum` are really stupid.

**Comment by kinke** [comment]
> > > Interesting ... wasn't -lowmem introduced because of an OOM on the auto-tester or was it a different CI?
> >
> > Yeah, very good point - f9a4e89
>
> It's still the case:

Yeah, but at that time I've also enabled parallel building of d_do_test unittests + normal executable, which isn't parallelized anymore.

**Comment by WalterBright** [comment]
The test failure here is:
```
Test fail_compilation/diag10327.d failed: 
expected:
----
fail_compilation/diag10327.d(11): Error: module `test10327` is in file 'imports/test10327.d' which cannot be read
import path[0] = fail_compilation
import path[1] = $p:druntime/import$
import path[2] = $p:phobos$
----
actual:
----
fail_compilation/diag10327.d(11): Error: module `test10327` is in file 'imports/test10327.d' which cannot be read
import path[0] = fail_compilation
import path[1] = /Users/braddr/sandbox/at-client/pull-4136150-Darwin_64_64/dmd/test/../../druntime/import
import path[2] = /Users/braddr/sandbox/at-client/pull-4136150-Darwin_64_64/dmd/test/../../phobos
----
diff:
----
 fail_compilation/diag10327.d(11): Error: module `test10327` is in file 'imports/test10327.d' which cannot be read
 import path[0] = fail_compilation
-import path[1] = $p:druntime/import$
-import path[2] = $p:phobos$
+import path[1] = /Users/braddr/sandbox/at-client/pull-4136150-Darwin_64_64/dmd/test/../../druntime/import
+import path[2] = /Users/braddr/sandbox/at-client/pull-4136150-Darwin_64_64/dmd/test/../../phobos
----
```

**Comment by MoonlightSentinel** [comment]
@wilzbach The main memory requirement stems from the CTFE done for regex. We could disable it without `-lowmem` using something like this:

```diff
diff --git a/test/tools/d_do_test.d b/test/tools/d_do_test.d
index aecb61173..14111c8c1 100755
--- a/test/tools/d_do_test.d
+++ b/test/tools/d_do_test.d
@@ -613,13 +613,14 @@ string unifyNewLine(string str)
     // On Windows, Outbuffer.writenl() puts `\r\n` into the buffer,
     // then fprintf() adds another `\r` when formatting the message.
     // This is why there's a match for `\r\r\n` in this regex.
-    static re = regex(`\r\r\n|\r\n|\r|\n`, "g");
+    static re = buildRegex(`\r\r\n|\r\n|\r|\n`, "g");
     return std.regex.replace(str, re, "\n");
 }
 
 string unifyDirSep(string str, string sep)
 {
-    static re = regex(`(?<=[-\w{}][-\w{}]*)/(?=[-\w][-\w/]*\.(di?|mixin)\b)`, "g");
+    static re = buildRegex(`(?<=[-\w{}][-\w{}]*)/(?=[-\w][-\w/]*\.(di?|mixin)\b)`, "g");
+
     return std.regex.replace(str, re, sep);
 }
 unittest
@@ -1662,3 +1663,26 @@ void printCppSources (in const(char)[][] compiled)
         }
     }
 }
+
+version (LazyRegex)
+{
+    struct buildRegex
+    {
+        string pattern, flags;
+        typeof(regex(pattern, flags)) re = void;
+
+        auto ref getRegex()
+        {
+            if (pattern !is null)
+            {
+                re = regex(pattern, flags);
+                pattern = null;
+            }
+            return re;
+        }
+
+        alias getRegex this;
+    }
+}
+else
+    alias buildRegex = regex;
```

Don't think no-CTFE should be the default because it implies a startup penalty for every invocation of `d_do_test` which could stack up a lot given the amount of individual tests,

**Comment by wilzbach** [comment]
> We could disable it without -lowmem using something like this:

AFAICT the `unifyNewLine` regular expression isn't very memory expensive (and could be easily replaced with a normal replace function). The `unifyDirSep`, however, costs about half of the total 1.5GB, so I just wanted to see whether we can roll a more naive version as a function.

I like your approach too and it's definitely the a good fallback if we have to use the regex.
Another easy workaround would be to put `unifyNewLine` in a separate file, s.t. we can compile it separately and at least have the memory consumption halfed that way.

**Comment by wilzbach** [comment]
> The unifyDirSep, however, costs about half of the total 1.5GB, so I just wanted to see whether we can roll a more naive version as a function.

So it looks like this works for all test suites. It's less elegant than the regular expression, but reduces the memory consumption from 1.5G to 740M (without -lowmem). I kept it rather verbose, s.t. it's easier to understand, but if line-count is an issue, it could be condensed.

**Review by MoonlightSentinel — COMMENTED** [review]
Please use Phobos style to keep this code consistent.
Otherwise LGTM.

File: test/tools/d_do_test.d (hunk 3 above)
Comment (wilzbach) [reply test/tools/d_do_test.d:1013]: Hmm, it's related to this commit as conditional only made sense with `if (!conditional)` which was removed here.
Anyhow, I split the commit into two.

**Comment by wilzbach** [comment]
> Please use Phobos style to keep this code consistent.

I had another look at this and found a more elegant non-regex approach which should be good enough for this use. Let's see what the CIs say.

**Comment by MoonlightSentinel** [comment]
Nice but not enough:

```diff
-compilable\ddoc_markdown_links_verbose.d(28): Ddoc: found link reference 'dub' to 'https:\\code.dlang.org'
+compilable\ddoc_markdown_links_verbose.d(28): Ddoc: found link reference 'dub' to 'https://code.dlang.org'
```

Maybe add a lookahead to check that the suffix is an actual file extension?

**Comment by wilzbach** [comment]
> Maybe add a lookahead to check that the suffix is an actual file extension?

Added. Looks like the CIs are happy with it (:

**Comment by wilzbach** [comment]
CC @thewilsonator. This is ready and as the d_do_test segfaults keep happening randomly, maybe this can get a bit of priority?

**Review by thewilsonator — APPROVED** [review]

**Review by MoonlightSentinel — APPROVED** [review]
To get the CI working for other PR's.

Not sure if there are other edge cases for `unifyDirSep` but it should work for dmd's error messages AFAICT.
